internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+
+	// Do all the copying from the BMH annotations to the agent.
+
+	annotations := bmh.ObjectMeta.GetAnnotations()
+	if val, ok := annotations[BMH_AGENT_ROLE]; ok {
+		switch val {
+		case "master":
+			agent.Spec.Role = models.HostRoleMaster
+		case "worker":
+			agent.Spec.Role = models.HostRoleWorker
+			// should we do something?
+			// default:
+		}
+	}
+
+	if val, ok := annotations[BMH_AGENT_HOSTNAME]; ok {
+		agent.Spec.Hostname = val
+	}
+
+	if val, ok := annotations[BMH_AGENT_MACHINE_CONFIG_POOL]; ok {
+		agent.Spec.MachineConfigPool = val
+	}
+
+	agent.Spec.Approved = true
+	if agent.ObjectMeta.Labels == nil {
+		agent.ObjectMeta.Labels = make(map[string]string)
+	}
+
+	// Label the agent with the reference to this BMH
+	agent.ObjectMeta.Labels[AGENT_BMH_LABEL] = bmh.Name
+
+	// TODO: Analyze the root device hints
+	return reconcileComplete{true}
+}
+
+// Reconcile BMH's HardwareDetails using the agent's inventory
+//
+// Here we will copy as much data from the Agent's Inventory as possible
+// into the `inspect.metal3.io/hardwaredetails` annotation in the BMH.
+//
+// This will trigger a reconcile on the BMH side, resulting in this data
+// being copied from the annotation into the BMH's HardwareDetails status.
+//
+//
+// Care must be taken to only update the data when really needed. Dong an update
+// on every BMAC reconcile will trigger an infinite loop of reconciles between
+// BMAC and the BMH reconcile as the former will update the hardwaredetails annotation
+// while the latter will continue to update the status.
+func (r *BMACReconciler) reconcileAgentInventory(bmh *bmh_v1alpha1.BareMetalHost, agent *adiiov1alpha1.Agent) reconcileResult {
+	// This check should be updated. We should check the
+	// agent status instead.
+	if len(agent.Status.Inventory.Interfaces) == 0 {
+		return reconcileComplete{}
+	}
+
+	bmhAnnotations := bmh.ObjectMeta.GetAnnotations()
+	if _, ok := bmhAnnotations[BMH_HARDWARE_DETAILS_ANNOTATION]; ok {
+		return reconcileComplete{}
+	}
+
+	// Check whether HardwareDetails has been set already. This annotation
+	// status should only be set through this Reconcile in this scenario.
+	if bmh.Status.HardwareDetails != nil {
+		return reconcileComplete{}
+	}
+
+	inventory := agent.Status.Inventory
+	hardwareDetails := bmh_v1alpha1.HardwareDetails{}
+
+	// Add the interfaces
+	for _, iface := range inventory.Interfaces {
+		// BMH handles dual stack in a different way, it requires adding
+		// multiple NICs, same mac and name, with different IPs.
+		// reference: https://github.com/metal3-io/baremetal-operator/pull/758
+		for _, ip := range append(iface.IPV6Addresses, iface.IPV4Addresses...) {
+			hardwareDetails.NIC = append(hardwareDetails.NIC, bmh_v1alpha1.NIC{
+				IP:        ip,
+				Name:      iface.Name,
+				Model:     iface.Vendor,
+				MAC:       iface.MacAddress,
+				SpeedGbps: int(iface.SpeedMbps),
```

Comment:
Units mismatch?

Reply:
:facepalm: good catch. I forgot to do the conversion after rewriting this part.